pytest/tests/sanity/upgradable.py

```diff
@@ -48,15 +49,22 @@ def get_proto_version(exe: pathlib.Path) -> int:
                    f'Got {line.rstrip()} on standard output')
         return m.group(1), int(m.group(2))
 
-    stable_release, stable_proto = get_proto_version(executables.stable.neard)
-    curr_release, curr_proto = get_proto_version(executables.current.neard)
-    assert stable_proto == curr_proto or stable_proto + 1 == curr_proto, (
-        f'If changed, protocol version of a new release can increase by at '
-        'most one.\n'
-        f'Got protocol {stable_proto} version in release {stable_release}\n'
-        f'but protocol {curr_proto} version in {curr_release}')
-    logger.info(f'Got proto {stable_proto} in release {stable_release} and '
-                f'and proto {curr_proto} in {curr_release}')
+    main_release, main_proto = get_proto_version(executables.stable.neard)
+    test_release, test_proto = get_proto_version(testnet.neard)
+    _, head_proto = get_proto_version(executables.current.neard)
+
+    lines = (
+        f'Got protocol {main_proto} in release {main_release} on mainnet.',
+        f'Got protocol {test_proto} in release {test_release} on testnet.',
+        f'Got protocol {head_proto} on master branch.',
+    )
+
+    ok = (head_proto in (test_proto, test_proto + 1) and
+          test_proto in (main_proto, main_proto + 1))
```

Comment:
:+1: great that we now testing this directly!

Reply:
> why the zero-hypothesis of "let's just not bump master until mainnet release" doesn't work? As in, why do we need this PR?

With the current release schedule I’m not sure if there’s a period between mainnet release and next release cut.  We are now going to implement cut happening a week before testnet release (so we have a week for testing) and that will coincide with when mainnet release happens.

Furthermore, it introduces disconnect between when a protocol-changing feature is committed and when when the version is increased.  It’s not unreasonable to think that someone might forget to bump the version if they need to remember to do it weeks after they’ve landed their initial change.

> One thing which feels a bit suboptimal is that the new scheme prevents us from releasing directly from master.

It doesn’t.  You just need to be mindful of any protocol changes such release might introduce.  If anything, this makes it safer to make releases directly from master since it will prevent you from releasing without noticing that there are new protocol changes.